cs/src/core/Index/FasterLog/FasterLog.cs

```diff
@@ -179,6 +180,40 @@ public unsafe bool TryAppend(Span<byte> entry, ref long logicalAddress)
             return true;
         }
 
+        private TaskCompletionSource<long> commitTask = new TaskCompletionSource<long>();
+
+        /// <summary>
+        /// Append entry to log (async) - completes after entry is flushed to storage
+        /// </summary>
+        /// <param name="entry"></param>
+        /// <returns></returns>
+        public async ValueTask<long> AppendAsync(byte[] entry)
+        {
+            long logicalAddress = 0;
+
+            while (true)
+            {
+                if (TryAppend(entry, ref logicalAddress))
+                    break;
+
+                await commitTask.Task;
```

Comment:
since appends are concurrent, any need to protect this commitTask access with a lock / interlocked.CompareExchange ? We don't want to accidentally use a stale commitTask that's already in completed state due to a past commit ?

Reply:
We order the memory access so that the task is cached, then the commit address is read. Since the committer first updates commit address, then updates the task, we should be fine. Using a stale task is fine because all state tasks will have a completed status, which will cause this loop to retry.